File: backend/app/schemas/organization.py
from __future__ import annotations

from datetime import datetime
from uuid import UUID

from pydantic import BaseModel, Field


class OrgUnitCreate(BaseModel):
    name: str = Field(min_length=1, max_length=255)
    unit_type: str = "TEAM"
    parent_id: UUID | None = None
    sort_order: int = 0
    competency_codes: list[str] = Field(default_factory=list)
    active_dimension_codes: list[str] | None = None


class OrgUnitUpdate(BaseModel):
    name: str | None = Field(default=None, max_length=255)
    unit_type: str | None = None
    parent_id: UUID | None = None
    sort_order: int | None = None
    competency_codes: list[str] | None = None
    active_dimension_codes: list[str] | None = None


class OrgUnitOut(BaseModel):
    id: UUID
    org_id: UUID
    parent_id: UUID | None
    name: str
    unit_type: str
    sort_order: int
    competency_codes: list[str]
    active_dimension_codes: list[str] | None = None
    children: list[OrgUnitOut] = []

    model_config = {"from_attributes": True}


OrgUnitOut.model_rebuild()


class OrganizationCreate(BaseModel):
    name: str = Field(min_length=1, max_length=255)
    industry: str | None = Field(default=None, max_length=100)


class OrganizationUpdate(BaseModel):
    name: str | None = Field(default=None, max_length=255)
    industry: str | None = Field(default=None, max_length=100)


class OrganizationOut(BaseModel):
    id: UUID
    tenant_id: UUID
    name: str
    industry: str | None
    created_at: datetime
    updated_at: datetime
    units: list[OrgUnitOut] = []

    model_config = {"from_attributes": True}


class OrganizationListOut(BaseModel):
    id: UUID
    name: str
    industry: str | None
    created_at: datetime
    unit_count: int = 0

    model_config = {"from_attributes": True}

File: frontend/types/organization.ts
export type UnitType = "BUSINESS_UNIT" | "DEPARTMENT" | "TEAM";

export interface OrgUnit {
  id: string;
  org_id: string;
  parent_id: string | null;
  name: string;
  unit_type: UnitType;
  sort_order: number;
  competency_codes: string[];
  active_dimension_codes: string[] | null;
  primary_contact_id: string | null;
  children: OrgUnit[];
}

export interface Organization {
  id: string;
  tenant_id: string;
  name: string;
  industry: string | null;
  primary_contact_id: string | null;
  created_at: string;
  updated_at: string;
  units: OrgUnit[];
}

export interface OrganizationListItem {
  id: string;
  name: string;
  industry: string | null;
  created_at: string;
  unit_count: number;
}

export interface OrganizationCreate {
  name: string;
  industry?: string;
}

export interface OrganizationUpdate {
  name?: string;
  industry?: string;
  primary_contact_id?: string | null;
}

export interface OrgUnitCreate {
  name: string;
  unit_type: UnitType;
  parent_id?: string;
  sort_order?: number;
  competency_codes?: string[];
  active_dimension_codes?: string[] | null;
  primary_contact_id?: string | null;
}

export interface OrgUnitUpdate {
  name?: string;
  unit_type?: UnitType;
  parent_id?: string | null;
  sort_order?: number;
  competency_codes?: string[];
  active_dimension_codes?: string[] | null;
  primary_contact_id?: string | null;
}
